internal/k8s/converter.go

```diff
@@ -0,0 +1,75 @@
+// Copyright © 2020 VMware
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+	"reflect"
+
+	ingressroutev1 "github.com/projectcontour/contour/apis/contour/v1beta1"
+	projectcontour "github.com/projectcontour/contour/apis/projectcontour/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// UnstructuredConverter handles conversions between unstructured.Unstructured and Contour types
+type UnstructuredConverter struct {
+	// scheme holds an initializer for converting Unstructured to a type
+	scheme *runtime.Scheme
+}
+
+// NewUnstructuredConverter returns a new UnstructuredConverter initialized
+func NewUnstructuredConverter() *UnstructuredConverter {
+	uc := &UnstructuredConverter{
+		scheme: runtime.NewScheme(),
+	}
+
+	// Setup converter to understand custom CRD types
+	projectcontour.AddKnownTypes(uc.scheme)
+	ingressroutev1.AddKnownTypes(uc.scheme)
+
+	return uc
+}
+
+// Convert converts an unstructured.Unstructured to typed struct
+// Any other type passed other than unstructured is just returned back to caller
+func (c *UnstructuredConverter) Convert(obj interface{}) (interface{}, error) {
+	switch obj := obj.(type) {
+	case *unstructured.Unstructured:
```

Comment:
I think this out switch can be rewritten as
```
unstructured, ok := obj.(*unstructured.Unstructured)
if !ok {
    return nil, fmt.Errorf("unsupported object type: %v", reflect.TypeOf(obj))
}
switch unstructured.GetKind() {
...
}
```
to reduce the overall indentation in this method

Reply:
This changes up the original idea of what this method did which I think is better. Before you could pass it a known type or an unstructured and it would return, but now it only converts an unstructured. This caused a check in the handler to only apply to the unstructured types.